content/posts/using-the-redis-allocator-in-rust.md

```diff
@@ -0,0 +1,151 @@
+6   libhello.dylib                      0x000000010af6ea3f _ZN4core9panicking9panic_fmt17h09741a3213dba543E + 63
+7   libhello.dylib                      0x000000010af6e984 _ZN4core9panicking5panic17hb4bc64e7f35c9151E + 100
+8   libhello.dylib                      0x000000010af53108 _ZN4core6option15Option$LT$T$GT$6unwrap17h66957b4d942a4d3cE + 56
+9   libhello.dylib                      0x000000010af4f8f3 _ZN76_$LT$redis_module..alloc..RedisAlloc$u20$as$u20$core..alloc..GlobalAlloc$GT$5alloc17h6588ea2d7520a3ebE + 35
+...
+```
+
+So, it looks like we had a null pointer dereference here (`3 ??? 0x0000000000000000 0x0 + 0`), but what are all these weird symbols starting with `_ZN...`?
+
+A bit of searching turns up that this is the way Rust does name mangling: Unlike in C, and similarly to C++, in Rust multiple functions with the same name can coexist, since there are various namespace mechanisms such as modules and traits to distinguish them. To generate unique symbols that are C-compatible, the compiler mangles these to long and ugly unique names. To demangle these names back into the original, we can filter the output through [`rustfilt`](https://crates.io/crates/rustfilt). 
+
+This gives us the following stack trace (uninteresting parts removed):
+
+```plaintext
+3   ???                                 0x0000000000000000 0x0 + 0
+...
+31  libhello.dylib                      0x000000010af6e984 core::panicking::panic + 100
+32  libhello.dylib                      0x000000010af53108 core::option::Option<T>::unwrap + 56
+33  libhello.dylib                      0x000000010af4f8f3 <redis_module::alloc::RedisAlloc as core::alloc::GlobalAlloc>::alloc + 35
+34  libhello.dylib                      0x000000010af4cc8c __rg_alloc + 60
+35  libhello.dylib                      0x000000010af6e2f6 <alloc::vec::Vec<u8> as core::convert::From<&str>>::from + 38
+36  libhello.dylib                      0x000000010af4de54 <T as core::convert::Into<U>>::into + 36
+37  libhello.dylib                      0x000000010af4f57f std::ffi::c_str::CString::new + 47
+38  libhello.dylib                      0x000000010af40daa RedisModule_OnLoad + 58
+39  redis-server                        0x000000010adf97d9 moduleLoad + 118
+40  redis-server                        0x000000010adf9735 moduleLoadFromQueue + 69
+41  redis-server                        0x000000010ad94428 main + 1190
+...
+```
+
+It still took me a lot head scratching and experimenting to figure it out, but here's what happened:
+
+As the stack trace shows, during the loading of the module we call the `CString::new` function. This standard library library function allocates memory for a string. This, in turn, calls our allocator which would then call `RedisModule_Alloc.unwrap()...` to actually perform the allocation. This causes a chicken-and-egg problem. The Redis module is not ready yet, meaning our function pointers have not yet been initialized, so we can't call the relevant API to perform the allocation.
```

Comment:
It might be worth spending a few words on why the module being not ready causes the func pointers to be null. People that don't know abut the internals of redismodule.h might not expect that there's an initialization step that was "snuck" in by antirez.

Reply:
Good idea, will do.

Reply:
@kristoff-it I clarified this a bit. I can't say I fully understand the reason why this custom initialization exists, but I hope the explanation is clearer now.